service_provider.go

```diff
@@ -406,18 +410,35 @@ func (sp *ServiceProvider) ParseResponse(req *http.Request, possibleRequestIDs [
 		return nil, requestID, retErr
 	}
 
-	if resp.Destination != sp.AcsURL.String() {
+	// AUTH-2414: If the message is signed, the Destination XML attribute in the root SAML element of the protocol
+	// message MUST contain the URL to which the sender has instructed the user agent to deliver the
+	// message. The recipient MUST then verify that the value matches the location at which the message has
+	// been received. Ref: http://docs.oasis-open.org/security/saml/v2.0/saml-bindings-2.0-os.pdf @ 3.4.5.2
+	//
+	// Unfortunately, PingFederate does not send Destination, but Okta and OneLogin do.
+	// This will check Destination only if provided. The Recipient is checked to the ACSURL in validateAssertion()
+	// so this particular check is not terribly important.
+	if resp.Destination != "" && resp.Destination != sp.AcsURL.String() {
 		retErr.PrivateErr = fmt.Errorf("`Destination` does not match AcsURL (expected %q)", sp.AcsURL.String())
 		return nil, requestID, retErr
 	}
 
 	requestIDvalid := false
-	for _, possibleRequestID := range possibleRequestIDs {
-		if resp.InResponseTo == possibleRequestID {
-			requestIDvalid = true
-			requestID = possibleRequestID
+
+	// AUTH-2414: This will allow IDP-Initiated requests to succeed if possibleRequestIDs is empty.
+	// This is a typical case with PingFederate.
+	// Ref: https://github.com/crewjam/saml/issues/151#issuecomment-435626671
+	if sp.AllowIDPInitiated {
```

Comment:
Given this check it would not check the `InResponseTo` for all SAML transactions regardless of SP or IdP initiated if `AllowIDPInitiated` is **true**. 

I would want to remove the if test here, run the for-loop and then have an if block that checks only for `requestIDvalid == false && sp.AllowIDPInitiated == true` to then flip the value for the Ping Federate behavior.

Reply:
I have moved the check for `AllowIDPInitiated` after the request ID validation has been done and made it explicit.